src/rpc/mining.cpp

```diff
@@ -578,24 +585,36 @@ UniValue getblocktemplate(const UniValue& params, bool fHelp)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
-        LEAVE_CRITICAL_SECTION(cs_main);
         {
-            checktxtime = boost::get_system_time() + boost::posix_time::minutes(1);
+            checktxtime = boost::get_system_time() + boost::posix_time::seconds(10);
 
+            int nHeight = chainActive.Tip()->nHeight;
             boost::unique_lock<boost::mutex> lock(csBestBlock);
             while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
             {
-                if (!cvBlockChange.timed_wait(lock, checktxtime))
-                {
-                    // Timeout: Check transactions for update
-                    if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
-                        break;
-                    checktxtime += boost::posix_time::seconds(10);
+                // Before waiting, generate the coinbase for the block following the next
+                // block (since this is cpu-intensive), so that when next block arrives,
+                // we can quickly respond with the template for following block.
+                if (!next_coinbase_mtx && IsShieldedMinerAddress(minerAddress)) {
+                    next_coinbase_mtx = CreateCoinbaseTransaction(Params(), CAmount{0}, minerAddress, nHeight + 2);
                 }
+                // Release the main lock while waiting
+                LEAVE_CRITICAL_SECTION(cs_main);
+                bool timedout = !cvBlockChange.timed_wait(lock, checktxtime);
+                ENTER_CRITICAL_SECTION(cs_main);
+
+                checktxtime += boost::posix_time::seconds(10);
```

Comment:
Previously, `checktxtime` would only be incremented if there were a timeout. Now, it is incremented even if `cvBlockChange` reported a block tip change.